xw/src/main/scala/uk/ac/surrey/xw/extension/prim/OnChange.scala

```diff
@@ -0,0 +1,86 @@
+package uk.ac.surrey.xw.extension.prim
+
+import org.nlogo.api.{Argument, Context, DefaultCommand}
+import org.nlogo.api.Syntax.{StringType, commandSyntax, CommandBlockType, CommandTaskType}
+import org.nlogo.nvm.{Context => NvmContext, _}
+import org.nlogo.workspace.AbstractWorkspace
+import uk.ac.surrey.xw.api.{PropertyKey, WidgetKey}
+import uk.ac.surrey.xw.extension.{KindInfo, WidgetContextManager}
+import uk.ac.surrey.xw.state.{SetProperty => SetPropEvent, RemoveWidget, StateEvent, Writer}
+
+import scala.collection.mutable.{Publisher, Subscriber}
+import scala.collection.parallel.mutable.ParMap
+
+case class ChangeListener(func: StateEvent => Unit)  extends Subscriber[StateEvent, Publisher[StateEvent]] {
+  def notify(pub: Publisher[StateEvent], event: StateEvent): Unit = func(event)
+}
+
+object OnChange {
+  val listeners = ParMap.empty[(WidgetKey, PropertyKey), Seq[ChangeListener]]
```

Comment:
It's a shame that `scala.collection.mutable.Publisher` doesn't expose its subscribers at all. You wouldn't need that if it did.